stdlib/Pkg/src/API.jl

```diff
@@ -547,4 +548,31 @@ function instantiate(ctx::Context; manifest::Union{Bool, Nothing}=nothing, kwarg
     Operations.build_versions(ctx, union(new_apply, new_git))
 end
 
+const ACTIVE_ENV = Ref{Union{String,Nothing}}(nothing)
+
+function _activate(env::Union{String,Nothing})
+    if env === nothing
+        @warn "Current directory is not in a project, nothing activated."
+    else
+        if !isempty(LOAD_PATH) && ACTIVE_ENV[] === LOAD_PATH[1]
+            LOAD_PATH[1] = env
+        else
+            # TODO: warn if ACTIVE_ENV !== nothing ?
```

Comment:
This could happen if, e.g. the user had run `insert!(LOAD_PATH, 1, new_path)`, thus causing `ACTIVE_ENV` to be out of sync with `LOAD_PATH`?

If that is true, I don't think we should do this shadowing of `LOAD_PATH`, I think we need to integrate `ACTIVE_ENV` into the code loading machinery.  E.g., when running `using Foo` we should _always_ search `vcat(ACTIVE_ENV, LOAD_PATH...)` instead of just `LOAD_PATH`.  This so that it is impossible to get into a "weird" situation where `ACTIVE_ENV` is actually shadowed by something that has been pushed onto the front of `LOAD_PATH`.  The separation of concerns is then a little more clear between `ACTIVE_ENV` and `LOAD_PATH`:

* `ACTIVE_ENV`: Used to set the "immediate" context of package loading and package manipulation.  Most users will only ever need to use this, and don't need to bother with `LOAD_PATH`.  Always takes precedence over anything else.

* `LOAD_PATH`: Used to set the "long tail" of search directories.  Used to fall back to system-provided packages, bundled goodies through some large Julia distribution, the `stdlib`, etc...

Additionally, this would simplify the `JULIA_LOAD_PATH` and `"@"`/`"@@"` business; we could instead define all those behaviors in reference to `ACTIVE_ENV`; `LOAD_PATH` can remain a dumb list that just has a single element popped onto its front if there is an active environment.

What do you think?

Reply:
I wrote the following in another place:

> A `JULIA_ENV` being set / unset by `pkg> activate dir` `pkg> deactivate` respectively and packages are searched in `[JULIA_ENV; LOAD_PATH]` (like now). Package manager also search for a project in `[JULIA_ENV; LOAD_PATH]` for which is the active project that commands are applied to.

which I think is similar to what @staticfloat describes. At the same time, I realize it is convenient to only have one list to search through things. I'll try make a PR towards the design above and we can see how it feels.

Reply:
Here is an example: https://github.com/JuliaLang/julia/commit/676df7bd3a8ea3182ae0ffe07b6f89a47b8c8297?w=1